src/mds/Server.cc

```diff
@@ -6450,6 +6450,11 @@ int Server::xattr_validate(CInode *cur, const InodeStoreBase::xattr_map_const_pt
       return -CEPHFS_ENODATA;
     }
 
+    if ((flags & CEPH_XATTR_REMOVE) && !(xattrs && xattrs->count(mempool::mds_co::string(xattr_name)))) {
```

Comment:
Yes, it will. For kclient when removing the ACLs it will send `setattr` requests only with `CEPH_XATTR_REMOVE` flags set.

More detail please see:

```
ret = __ceph_setxattr(inode, name, value, size, 0);
```

In https://github.com/ceph/ceph-client/blob/for-linus/fs/ceph/acl.c#L90-L167

Please note the `value` will be `NULL` and `size` will be `0`, because in `fs/posix_acl.c` it will set the `acl` to null when calling the `inode->i_op->set_acl()`:

https://github.com/ceph/ceph-client/blob/for-linus/fs/posix_acl.c#L936-L958.

And finally in `ceph_sync_setxattr()` it will send a `setattr` request by setting the `CEPH_XATTR_REMOVE` flags set:

https://github.com/ceph/ceph-client/blob/for-linus/fs/ceph/xattr.c#L1121-L1135.

Reply:
OK, then for the posix acts, the client would catch -ENODATA and treat it as success as done in #55087 ?

Reply:
Yeah, you just remind me that we should make it work with old kclients.

Just in case for old kclient without https://patchwork.kernel.org/project/ceph-devel/list/?series=832760&archive=both, with this PR it will fail with `-ENODTAT`.

I will update both this PR and kclient patch.

Reply:
Just for completeness of this discussion (for anyone following), see - https://tracker.ceph.com/issues/64679#note-4

Reply:
@vshankar BTW, is that reproduceable always. The old is buggy, because for sync request it will return `-ENODATA` while when the `Fx` is issued it will return success.

Reply:
I left a note in the tracker @lxbsz

Reply:
> Yeah, you just remind me that we should make it work with old kclients.
> 
> Just in case for old kclient without https://patchwork.kernel.org/project/ceph-devel/list/?series=832760&archive=both, with this PR it will fail with `-ENODTAT`.
> 
> I will update both this PR and kclient patch.

I have updated the PR by adding a new flags `CEPH_XATTR_REMOVE2`. Then we can make sure it will be compatible with old clients.